server/application-server/src/main/java/de/tum/in/www1/hephaestus/gitprovider/pullrequest/github/GitHubPullRequestSyncService.java

```diff
@@ -0,0 +1,251 @@
+                    .filter(pullRequest -> {
+                        try {
+                            return sinceDate.isEmpty() || pullRequest.getUpdatedAt().after(sinceDate.get());
+                        } catch (IOException e) {
+                            logger.error("Failed to filter pull request {}: {}", pullRequest.getId(), e.getMessage());
+                            return false;
+                        }
+                    })
+                    .toList();
+
+            pullRequests.addAll(keepPullRequests);
+            if (keepPullRequests.size() != ghPullRequests.size()) {
+                break;
+            }
+        }
+
+        pullRequests.forEach(this::processPullRequest);
+        return pullRequests;
+    }
+
+    /**
+     * Processes a single GitHub pull request by updating or creating it in the
+     * local repository.
+     * Manages associations with repositories, labels, milestones, authors,
+     * assignees, merged by users,
+     * and requested reviewers.
+     *
+     * @param ghPullRequest the GitHub pull request to process
+     * @return the updated or newly created PullRequest entity, or {@code null} if
+     *         an error occurred
+     */
+    @Transactional
+    public PullRequest processPullRequest(GHPullRequest ghPullRequest) {
+        var result = pullRequestRepository.findById(ghPullRequest.getId())
+                .map(pullRequest -> {
+                    try {
+                        if (pullRequest.getUpdatedAt() == null || pullRequest.getUpdatedAt()
+                                .isBefore(DateUtil.convertToOffsetDateTime(ghPullRequest.getUpdatedAt()))) {
+                            return pullRequestConverter.update(ghPullRequest, pullRequest);
+                        }
+                        return pullRequest;
+                    } catch (IOException e) {
+                        logger.error("Failed to update pull request {}: {}", ghPullRequest.getId(), e.getMessage());
+                        return null;
+                    }
+                }).orElseGet(() -> pullRequestConverter.convert(ghPullRequest));
+
+        if (result == null) {
+            return null;
+        }
+
+        // Link with existing repository if not already linked
+        if (result.getRepository() == null) {
+            // Extract name with owner from the repository URL
+            // Example: https://api.github.com/repos/ls1intum/Artemis/pulls/9463
+            var nameWithOwner = ghPullRequest.getUrl().toString().split("/repos/")[1].split("/pulls")[0];
+            var repository = repositoryRepository.findByNameWithOwner(nameWithOwner);
+            if (repository != null) {
+                result.setRepository(repository);
+            }
+        }
+
+        // Link new labels and remove labels that are not present anymore
+        var ghLabels = ghPullRequest.getLabels();
+        var resultLabels = new HashSet<>(result.getLabels());
```

Comment:
To prevent old labels not being removed after an `unlabeled`-action, we can just stop reassigning the old ones:
```suggestion
        var resultLabels = new HashSet<Label>();
```
Note: also requires a Label-import at the top:
`import de.tum.in.www1.hephaestus.gitprovider.label.Label;`